python/pyspark/pandas/series.py

```diff
@@ -4475,6 +4477,146 @@ def replace(
+        highest clocked speeds of different birds.
+
+        >>> from pyspark.pandas.config import set_option, reset_option
+        >>> set_option("compute.ops_on_diff_frames", True)
+        >>> s1 = ps.Series({'falcon': 330.0, 'eagle': 160.0})
+        >>> s1
+        falcon    330.0
+        eagle     160.0
+        dtype: float64
+        >>> s2 = ps.Series({'falcon': 345.0, 'eagle': 200.0, 'duck': 30.0})
+        >>> s2
+        falcon    345.0
+        eagle     200.0
+        duck       30.0
+        dtype: float64
+
+        Now, to combine the two datasets and view the highest speeds
+        of the birds across the two datasets
+
+        >>> s1.combine(s2, max, return_type="double")
+        duck        NaN
+        eagle     200.0
+        falcon    345.0
+        dtype: float64
+
+        In the previous example, the resulting value for duck is missing,
+        because the maximum of a NaN and a float is a NaN.
+        So, in the example, we set ``fill_value=0``,
+        so the maximum value returned will be the value from some dataset.
+
+        >>> s1.combine(s2, max, fill_value=0, return_type="double")
+        duck       30.0
+        eagle     200.0
+        falcon    345.0
+        dtype: float64
+        >>> reset_option("compute.ops_on_diff_frames")
+        """
+        if not isinstance(other, Series) and not np.isscalar(other):
+            raise TypeError("unsupported type: %s" % type(other))
+
+        tmp_other_col = verify_temp_column_name(self._internal.spark_frame, "__tmp_other_col__")
+        if np.isscalar(other):
+            combined = DataFrame(
+                InternalFrame(
+                    spark_frame=self._internal.spark_frame.select(
+                        *self._internal.spark_columns
+                    ).withColumn(tmp_other_col, SF.lit(other)),
+                    index_spark_columns=self._internal.index_spark_columns,
+                )
+            )
+        elif same_anchor(self, other):
+            combined = self._psdf[self._column_label, other._column_label]
+        elif fill_value is None:
+            combined = combine_frames(self.to_frame(), other.to_frame())
+        else:
+            combined = self._combine_frame_with_fill_value(other, fill_value=fill_value)
+
+        @pandas_udf(returnType=return_type)  # type: ignore
+        def wrapped_func(x: pd.Series, y: pd.Series) -> pd.Series:
+            return x.combine(y, func)
+
+        internal = InternalFrame(
+            spark_frame=combined._internal.spark_frame.select(
+                *combined._internal.index_spark_columns,
+                wrapped_func(*combined._internal.data_spark_columns),
+                NATURAL_ORDER_COLUMN_NAME
+            ),
+            index_spark_columns=combined._internal.index_spark_columns,
+            column_labels=self._internal.column_labels if np.isscalar(other) else [None],
+            column_label_names=[None],
+        )
```

Comment:
We can create the new `InternalFrame` by leveraging `self._internal.with_new_sdf` to keep the metadata from `self`.

How about

```python
        scol = wrapped_func(*combined._internal.data_spark_columns)
        combined_sdf = combined._internal.spark_frame.select(
            *combined._internal.index_spark_columns,
            scol.alias(self._internal.spark_column_name_for(self.spark.column)),
            NATURAL_ORDER_COLUMN_NAME
        )
        internal = self._internal.with_new_sdf(combined_sdf)
```

?

Reply:
updated

Reply:
After I try to fix the failed test, I think we should not use ```self._internal.with_new_sdf``` because self and output have difference metadata (index_fields, data_fields). Ex
``` python
>>> psdf = ps.DataFrame({"s1": [1, 2, 3], "s2": [4, 5, 6]})
>>> def true_div(s1: int, s2: int) -> float:
>>>     return s1 / s2
>>> psdf["s1"].combine(psdf["s2"], true_div)
```
self has:
``` scala
index_fields = [InternalField(dtype=object,struct_field=StructField(__index_level_0__,StringType,false))]
data_fields = [InternalField(dtype=int64,struct_field=StructField(s1,LongType,false))]
```

output:
``` scala
index_fields = [InternalField(dtype=object,struct_field=StructField(__index_level_0__,StringType,true))]
data_fields = [InternalField(dtype=float64,struct_field=StructField(s1,DoubleType,true))]
```

Reply:
Oh, yeah it's my mistake.

I think we should get a return type from the function, not `self`.

Let's try the code below

```python
        _, return_type = pandas_on_spark_type(infer_return_type(func))

        @pandas_udf(returnType=return_type)  # type: ignore
        def wrapped_func(x: pd.Series, y: pd.Series) -> pd.Series:
            return x.combine(y, func)
```

Reply:
many thanks, updated. Please take another look when you have time.